File: android/proj.android/src/cn/htgames/doudizhu/NativeInterface.java
package cn.htgames.doudizhu;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Handler;
import android.os.Message;
import android.os.Vibrator;
import android.util.Log;
import android.widget.FrameLayout;

import com.aliencoder.TWebView;
import com.aliencoder.imagepicker.IImagePickerDelegate;
import com.aliencoder.imagepicker.ImagePicker;
import com.aliencoder.networking.INetWorkingDelegate;
import com.aliencoder.networking.TNetWorkingManager;
import com.aliencoder.tools.Utils;
import com.aliencoder.um.UMXinPingTai;
import com.igexin.sdk.PushManager;
import com.tencent.mm.sdk.modelmsg.SendMessageToWX;
import com.thirdplatform.wechat.WeChatManager;

import org.egret.egretframeworknative.engine.EgretGameEngine;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import cn.htgames.doudizhu.authenticate.AuthenticateDelegate;
import cn.htgames.doudizhu.authenticate.AuthenticateManager;
import cn.htgames.doudizhu.authenticate.BindDelegate;

/**
 * Created by rockyl on 16/9/18.
 */
public class NativeInterface implements AuthenticateDelegate, BindDelegate {
	protected static final String TAG = "NativeInterface";

	private interface IRuntimeInterface {
		public void callback(String message);
		// 因为遗留问题 callBack 也是接受的
	}

	Activity activity;
	EgretGameEngine gameEngine;
	FrameLayout rootView;

	TWebView webView;
	Vibrator vibrator;
	TNetWorkingManager manager;
	HashMap<String, JSONObject> callbacks;
	ImagePicker imagePicker;

	String authCallId, bindCallId;

	AuthenticateManager authenticateManager;
	
	CallbackHandler mRechargeHandler = null;
	CallbackHandler mShareHandler = null;

	public NativeInterface(Activity activity, EgretGameEngine gameEngine, FrameLayout rootView){
		this.activity = activity;
		this.gameEngine = gameEngine;
		this.rootView = rootView;

		setup();
	}

	private void setup() {
		authenticateManager = new AuthenticateManager(activity, rootView);
		authenticateManager.authenticateDelegate = this;
		authenticateManager.bindDelegate = this;
		vibrator = (Vibrator)activity.getSystemService(Context.VIBRATOR_SERVICE);
		webView = new TWebView();
		manager = TNetWorkingManager.getInstance();
		callbacks = new HashMap<>();
		imagePicker = new ImagePicker(activity);

		// Egret（TypeScript）－Runtime（Java）通讯
		// setRuntimeInterface(String name, IRuntimeInterface interface) 用于设置一个runtime的目标接口
		// callEgretInterface(String name, String message) 用于调用Egret的接口，并传递消息
		gameEngine.setRuntimeInterface("egretCall", new IRuntimeInterface() {
			@Override
			public void callback(String message) {
				Log.d(TAG, message);

				try {
					JSONObject params = new JSONObject(message);
					String callId = params.getString("id");
					String method = params.getString("method");
					JSONObject args = params.getJSONObject("args");

					callbacks.put(callId, params);
					
					JSONObject retArgs = null;
					switch (method) {
						case "getEnvInfo":            //获取环境信息
							retArgs = getEnvInfo(args, callId);
							break;
						case "userLogin":             //当玩家登陆成功时
							retArgs = userLogin(args, callId);
							break;
						case "userLogout":            //当玩家注销时
							retArgs = userLogout(args, callId);
							break;
						case "getDeviceUUID":         //获取设备唯一码
							retArgs = getDeviceUUID(args, callId);
							break;
						case "closeApp":              //关闭app
							retArgs = closeApp(args, callId);
							break;
						case "navigateToUrl":         //用浏览器打开网址
							retArgs = navigateToUrl(args, callId);
							break;
						case "uploadImageFromDevice": //上传图片
							retArgs = uploadImageFromDevice(args, callId);
							break;
						case "vibrate":               //震动
							retArgs = vibrate(args, callId);
							break;
						case "addWebView":            //打开一个浏览器视图
							retArgs = addWebView(args, callId);
							break;
						case "removeWebView":         //移除浏览器视图
							retArgs = removeWebView(args, callId);
							break;
						case "alert":                 //弹出原生提示框
							retArgs = alert(args, callId);
							break;
						case "authenticate":          //授权sdk
							retArgs = authenticate(args, callId);
							break;
						case "authThirdPart":         //直接调用第三方授权
							retArgs = authThirdPart(args, callId);
							break;
						case "bindThirdPart":         //第三方绑定
							retArgs = bindThirdPart(args, callId);
							break;
						case "recharge":              //充值
							retArgs = recharge(args, callId);
							break;
						case "UMLoginStatistics":
							UMLoginStatistics(args,callId);
							break;
						case "share":               //分享
							retArgs = share(args, callId);
							break;
						case "toClipboard":               //剪切板
							retArgs = toClipboard(args, callId);
							break;
					}

					if (retArgs != null) {
						returnCall(callId, retArgs);
					}

				} catch (JSONException e) {
					e.printStackTrace();
				}

			}
		});
	}

	private void returnCall(String callId, JSONObject retArgs) {
		JSONObject params = callbacks.remove(callId);

		String method = null;
		if (params!=null) {
			try {
				method = params.getString("method");
			} catch (JSONException e) {
				e.printStackTrace();
			}

			callEgret(callId, method, retArgs);
		}
	}

	public void callEgret(String callId, String method, JSONObject retArgs){
		JSONObject retObj = new JSONObject();
		try {
			retObj.put("id", callId);
			retObj.put("method", method);
			retObj.put("args", retArgs);
		} catch (JSONException e) {
			e.printStackTrace();
		}

		String retMessage = retObj.toString();

		gameEngine.callEgretInterface("nativeCall", retMessage);
	}

	private JSONObject getEnvInfo(JSONObject args, String callId) {
		JSONObject ret = new JSONObject();

		Environment env = Environment.getInstance();
		try {
			ret.put("channel_id", env.channelId);
		} catch (JSONException e) {
			e.printStackTrace();
		}

		return ret;
	}
	
	private JSONObject toClipboard(JSONObject args, String callId) {
		JSONObject ret = new JSONObject();

		try {
			ClipboardManager cm = (ClipboardManager)activity.getSystemService(Context.CLIPBOARD_SERVICE);
	        // 将文本内容放到系统剪贴板里。
	        cm.setPrimaryClip(ClipData.newPlainText(args.getString("text"), args.getString("text")));
		} catch (Exception e) {
			e.printStackTrace();
		}

		return ret;
	}

	private JSONObject userLogin(JSONObject args, String callId) {
		GameInfo.getInstance().update(args);

		String uid = GameInfo.getInstance().uid;
		
		// 顺手上传用户信息
		submitLoginInfo();

		boolean result = PushManager.getInstance().bindAlias(activity, uid);
		Log.i(TAG, "绑定结果：" + result);
		return null;
	}

	private JSONObject userLogout(JSONObject args, String callId) {
		boolean result = PushManager.getInstance().unBindAlias(activity, GameInfo.getInstance().uid, true);
		Log.i(TAG, "解绑结果：" + result);

		return null;
	}

	private JSONObject getDeviceUUID(JSONObject args, String callId) {
		JSONObject ret = new JSONObject();

		String uuid = Utils.getUUIDWithMD5(activity);

		try {
			ret.put("uuid", uuid);
		} catch (JSONException e) {
			e.printStackTrace();
		}

		return ret;
	}

	private JSONObject closeApp(JSONObject args, String callId) {
		gameEngine.game_engine_onStop();
		activity.finish();

		return null;
	}

	private JSONObject navigateToUrl(JSONObject args, String callId) {
		try {
			String url = args.getString("url");
			Utils.openWithBrower(activity, url);
		} catch (JSONException e) {
			e.printStackTrace();
		}

		return null;
	}

	private JSONObject uploadImageFromDevice(JSONObject args, final String callId) {
		try {
			String title = args.getString("title");
			final String uploadUrl = args.getString("uploadUrl");

			imagePicker.show(title, new IImagePickerDelegate() {
				JSONObject ret = new JSONObject();

				@Override
				public void onPickImage(Bitmap image) {
					ByteArrayOutputStream stream = new ByteArrayOutputStream();
					image.compress(Bitmap.CompressFormat.JPEG, 100, stream);

					try {
						manager.upload(uploadUrl, stream.toByteArray(), new INetWorkingDelegate() {
							@Override
							public void receive(Object response) {
								JSONObject data = (JSONObject) response;
								try {
									int code = data.getInt("code");
									switch (code) {
										case 0:
											String url = data.getJSONObject("data").getString("file_path");
											ret.put("result", 0);
											ret.put("url", url);
											break;
										default:
											ret.put("result", 3);
											break;
									}
								} catch (JSONException e) {
									e.printStackTrace();
								}
								returnCall(callId, ret);
							}
						});
					} catch (IOException e) {
						e.printStackTrace();
						try {
							ret.put("result", 2);
						} catch (JSONException e1) {
							e1.printStackTrace();
						}
						returnCall(callId, ret);
					}
				}

				@Override
				public void onCancelPick() {
					try {
						ret.put("result", 1);
					} catch (JSONException e) {
						e.printStackTrace();
					}
					returnCall(callId, ret);
				}
			});
		} catch (JSONException e) {
			e.printStackTrace();
		}

		return null;
	}

	private JSONObject vibrate(JSONObject args, final String callId) {
		vibrator.vibrate(600);
		return null;
	}

	private JSONObject addWebView(JSONObject args, final String callId) {
		String url = null;
		try {
			url = args.getString("url");
		} catch (JSONException e) {
			e.printStackTrace();
		}

		webView.addTo(activity, rootView);
		webView.loadByUrl(url);
		return null;
	}

	private JSONObject removeWebView(JSONObject args, final String callId) {
		webView.remove();
		return null;
	}

	private JSONObject alert(JSONObject args, final String callId) {
		String title = null, message = null;
		String btn0 = null, btn1 = null;
		try {
			title = args.getString("title");
			message = args.getString("message");
			JSONArray buttons = args.getJSONArray("buttons");

			btn0 = buttons.getString(0);
			btn1 = buttons.getString(1);
		} catch (JSONException e) {
			e.printStackTrace();
		}

		final JSONObject retArgs = new JSONObject();

		AlertDialog alert;
		alert = new AlertDialog.Builder(activity).setTitle(title).setMessage(message)
			.setPositiveButton(btn0, new DialogInterface.OnClickListener() {
				@Override
				public void onClick(DialogInterface dialog, int which) {
					try {
						retArgs.put("index", 0);
					} catch (JSONException e) {
						e.printStackTrace();
					}
					returnCall(callId, retArgs);
				}
			})
			.setNegativeButton(btn1, new DialogInterface.OnClickListener() {
				@Override
				public void onClick(DialogInterface dialog, int which) {
					try {
						retArgs.put("index", 1);
					} catch (JSONException e) {
						e.printStackTrace();
					}
					returnCall(callId, retArgs);
				}
			})
			.create();

		alert.show();

		return null;
	}

	private JSONObject authenticate(JSONObject args, final String callId){
		/*Authenticate.getInstance().show(activity, new Handler(){
			@Override
			public void handleMessage(Message msg) {
				super.handleMessage(msg);

				returnCall(callId, (JSONObject) msg.obj);
			}
		});*/
		int type=1;
		try {
			type = args.getInt("type");
		}
		catch (JSONException e) {
			e.printStackTrace();
		}
		authCallId = callId;
		if (type==1)
			authenticateManager.showLoginByInput();
		if (type==2)
		{
			WebService.getInstance().loginByGuest(
					Utils.getUUID(activity),
					authenticateManager.onLoginResultHandler
			);
		}
		if (type==3)
			authenticateManager.showLoginByMobile();


		return null;
	}

	private JSONObject authThirdPart(JSONObject args, final String callId){
		int type = 0;
		try {
			type = args.getInt("type");
		} catch (JSONException e) {
			e.printStackTrace();
		}

		switch (type){
			case 0: //mobile

				break;
			case 1: //weixin
//				WXApiManager.getInstance().doAuth(activity, authenticateManager.onWXAuthCallback);
				WeChatManager.getInstance().login(authenticateManager.onWXAuthCallback);
				break;
		}

		authCallId = callId;

		return null;
	}

	private JSONObject bindThirdPart(JSONObject args, final String callId){
		int type = 0;
		try {
			type = args.getInt("type");
		} catch (JSONException e) {
			e.printStackTrace();
		}

		switch (type){
			case 0: //mobile
				authenticateManager.showBindByInput();
				break;
			case 1: //weixin
//				WXApiManager.getInstance().doAuth(activity, authenticateManager.onWXBindCallback);
				WeChatManager.getInstance().login(authenticateManager.onWXBindCallback);
				break;
		}

		bindCallId = callId;

		return null;
	}

	class CallbackHandler extends Handler {
		public String mCallId;

		public CallbackHandler(String callId) {
			mCallId = callId;
		}
		
		@Override
		public void handleMessage(Message msg) {
			super.handleMessage(msg);
			JSONObject retArgs = new JSONObject();
			try {
				retArgs.put("code", msg.what);
			} catch (JSONException e) {
				e.printStackTrace();
			}
			Log.i(TAG, "handleMessage：" + mCallId);
			returnCall(mCallId, retArgs);
		}
	}
	
	private JSONObject recharge(JSONObject args, final String callId){
		String pid = "";

		mRechargeHandler = new CallbackHandler(callId);
		
		try {
			pid = args.getString("id");
		} catch (JSONException e) {
			e.printStackTrace();
		}

		WeChatManager.getInstance().pay(pid, GameInfo.getInstance().uid,
				GameInfo.getInstance().serverType, mRechargeHandler);

		return null;
	}
	
	private JSONObject share(JSONObject args, final String callId){
		String title = "";
		String url = "";
		String desc = "";
		int scene = 0;

		mShareHandler = new CallbackHandler(callId);
		
		try {
			String shareScene = args.getString("type");
			scene = shareScene.equalsIgnoreCase("wx_timeline") ? 
					SendMessageToWX.Req.WXSceneTimeline : SendMessageToWX.Req.WXSceneSession;
			JSONObject params = args.getJSONObject("params");
			url = params.getString("url");
			desc = params.getString("description");
			title = params.getString("title");
		} catch (JSONException e) {
			e.printStackTrace();
		}

		WeChatManager.getInstance().share(url, title, desc, scene, mShareHandler);

		return null;
	}

	private void UMLoginStatistics(JSONObject args, final String callId) {
		String uid;
		int type;
		try {
			uid = args.getString("uid");
			type = args.getInt("type");
			if (type==0)
				UMXinPingTai.getInstance().onProfileSignIn(uid);
			else if (type==1)
				UMXinPingTai.getInstance().onProfileSignIn("WX",uid);
		} catch (JSONException e) {
			e.printStackTrace();
		}
	}


	@Override
	public void onAuthenticateResult(JSONObject response) {
		returnCall(authCallId, response);
	}

	@Override
	public void onAuthenticateCancel() {
		JSONObject response = null;
		try {
			response = new JSONObject("{code: 1}");
		} catch (JSONException e) {
			e.printStackTrace();
		}
		returnCall(authCallId, response);
	}

	@Override
	public void onBindResult(JSONObject response) {
		returnCall(bindCallId, response);
	}

	@Override
	public void onBindCancel() {
		JSONObject response = null;
		try {
			response = new JSONObject("{code: 1}");
		} catch (JSONException e) {
			e.printStackTrace();
		}
		returnCall(bindCallId, response);
	}

	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		imagePicker.onActivityResult(requestCode, resultCode, data);
	}
	
	private void submitLoginInfo() {
		String userToken = GameInfo.getInstance().token;
		String userId = GameInfo.getInstance().uid;
		String osType = "android";
		String channelId = activity.getString(R.string.install_channel);
		
		// 直接发送用户信息给web接口做统计
		Map<String, String> paramMap = new HashMap<String, String>();
		paramMap.put("uid", userId);
		paramMap.put("token", userToken);
		paramMap.put("os_type", osType);
		paramMap.put("channel_id", channelId);
		
		Log.d(TAG, "submitLoginInfo token:" + userToken 
				+ " uid:" + userId + " os_type:" + osType
				+ " channel_id:" + channelId);
		
		WebService.getInstance().callApi("user", "channel", paramMap, new Handler() {
				@Override
				public void handleMessage(Message msg) {
					JSONObject resp = (JSONObject)msg.obj;
					if (resp == null) {
						Log.e(TAG, "submitLoginInfo got no response!");
						return ;
					}
					
					try {
						int code = resp.getInt("code");
		
						if(code == 0) {
							Log.d(TAG, "submitLoginInfo success!");
						} else {
							Log.e(TAG, "submitLoginInfo failed!" + code);
						}
					} catch (JSONException e) {
						e.printStackTrace();
					}
					
					super.handleMessage(msg);
				}
			}, "POST", null, null, null);
	}
}

File: ios/newplatfrom/NativeInterface.m
//
//  NativeInterface.m
//  doudizhuhero
//
//  Created by 劳琪峰 on 16/5/27.
//  Copyright © 2016年 egret. All rights reserved.
//

#import <AudioToolbox/AudioToolbox.h>
#import <AFNetworking/AFNetworking.h>
#import "NativeInterface.h"
#import "EgretRuntime.h"
#import "ImagePicker.h"
#import "WebView.h"
#import "WebService.h"
#import "Utils.h"
#import "GameInfo.h"
#import "GeTuiSdk.h"
#import "AppPay.h"
#include <UMMobClick/MobClick.h>
#include <UMMobClick/MobClickGameAnalytics.h>

//#import "AliPayManager.h"



@interface NativeInterface()
@property (nonatomic,strong) WebView *webView;
@end

@implementation NativeInterface

UIViewController *viewController;
NSMutableDictionary *callbacks;
ImagePicker *imagePicker;
AFHTTPSessionManager *manager;
NSString *authCallId, *bindCallId;
HTGSDK* sdk;
AppPay* appPay;

- (instancetype)initWithViewController:(UIViewController *)vc {
	if (self = [super init]) {
		viewController = vc;
		
		[self setup];
	}
	return self;
}

- (void)setup{
	callbacks = [NSMutableDictionary dictionaryWithCapacity:10];
	imagePicker = [[ImagePicker alloc]initWithRoot:viewController];
	sdk = [[HTGSDK alloc]initWithViewController:viewController withDelegate:self];
	appPay = [[AppPay alloc]init];
    NSDictionary *infoDictionary = [[NSBundle mainBundle] infoDictionary];
    NSString *wxappid = infoDictionary[@"Wxappid"];
	[[WXApiManager getInstance]setup:wxappid];
	self.webView = [WebView alloc];
    
    [Utils setVc:viewController];
	
	NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
	manager = [[AFHTTPSessionManager alloc] initWithSessionConfiguration:configuration];
	
	[[EgretRuntime getInstance] setRuntimeInterface:@"egretCall" block:^(NSString *message) {
		NSLog(@"%@", message);
		NSData *data = [message dataUsingEncoding:NSUTF8StringEncoding];
		NSError * error = nil;
		id paramsId = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
		NSDictionary *params = (NSDictionary *)paramsId;
		NSString *callId = [params valueForKey:@"id"];
		NSString *method = [params valueForKey:@"method"];
		NSDictionary *args = [params valueForKey:@"args"];
		
		NSDictionary *retArgs;
		if([method isEqualToString:@"getEnvInfo"]){
			retArgs = [self getEnvInfo:args callId:callId];
		}if([method isEqualToString:@"userLogin"]){
			retArgs = [self onUserLogin:args callId:callId];
		}if([method isEqualToString:@"userLogout"]){
			retArgs = [self onUserLogout:args callId:callId];
		}else if([method isEqualToString:@"getDeviceUUID"]){
			retArgs = [self getDeviceUUID:args callId:callId];
		}else if([method isEqualToString:@"navigateToUrl"]){
			retArgs = [self navigateToUrl:args callId:callId];
		}else if([method isEqualToString:@"uploadImageFromDevice"]){
			retArgs = [self uploadImageFromDevice:args callId:callId];
		}else if([method isEqualToString:@"vibrate"]){
			retArgs = [self vibrate:args callId:callId];
		}else if([method isEqualToString:@"share"]){
			retArgs = [self share:args callId:callId];
		}else if([method isEqualToString:@"alert"]){
			retArgs = [self alert:args callId:callId];
		}else if([method isEqualToString:@"addWebView"]){
			retArgs = [self addWebView:args callId:callId];
		}else if([method isEqualToString:@"removeWebView"]){
			retArgs = [self removeWebView:args callId:callId];
		}else if([method isEqualToString:@"authenticate"]){
			retArgs = [self authenticate:args callId:callId];
		}else if([method isEqualToString:@"authThirdPart"]){
			retArgs = [self authThirdPart:args callId:callId];
		}else if([method isEqualToString:@"bindThirdPart"]){
			retArgs = [self bindThirdPart:args callId:callId];
        }else if([method isEqualToString:@"recharge"]){
            retArgs = [self recharge:args callId:callId];
        }else if([method isEqualToString:@"UMLoginStatistics"]){
			retArgs = [self UMLoginStatistics:args callId:callId];
        }else if([method isEqualToString:@"toClipboard"]){
            retArgs = [self toClipboard:args callId:callId];
        }
		
		[callbacks setValue:params forKey:callId];
		if(retArgs){
			[self returnCall:callId retArgs:retArgs];
		}
	}];
}

- (void)returnCall:(NSString *)callId retArgs:(NSDictionary *)retArgs{
	NSDictionary *params = callbacks[callId];
	[callbacks removeObjectForKey:callId];
	NSString *method = [params valueForKey:@"method"];
	NSDictionary * retObj = [NSDictionary dictionaryWithObjectsAndKeys:
													 callId, @"id",
													 method, @"method",
													 retArgs, @"args",
													 nil];
	NSError * error = nil;
	NSData * data = [NSJSONSerialization dataWithJSONObject:retObj options:kNilOptions error:&error];
	NSString *retMessage = [[NSString alloc]initWithData:data encoding:NSUTF8StringEncoding];
	[[EgretRuntime getInstance] callEgretInterface:@"nativeCall" value:retMessage];
}

#pragma implements
- (NSDictionary *)getEnvInfo:(NSDictionary * )args callId:(NSString *)callId{
	NSDictionary *ret = [NSDictionary dictionaryWithObjectsAndKeys:
											 [NSNumber numberWithInteger:2], @"channel_id",
											 nil];
	
	return ret;
}

- (NSDictionary *)onUserLogin:(NSDictionary * )args callId:(NSString *)callId{
    NSLog(@"userLogin args = %@",args);
	[[GameInfo getInstance]update:args];
    
    [self submitLoginInfo];
	
	NSString *uid = [GameInfo getInstance].uid;
	NSLog(@"[GTSdk bindAlias] %@", uid);
	[GeTuiSdk bindAlias:uid andSequenceNum:@"1234567890"];
	
	return nil;
}

- (NSDictionary *)onUserLogout:(NSDictionary * )args callId:(NSString *)callId{
	NSString *uid = [GameInfo getInstance].uid;
	NSLog(@"[GTSdk unbindAlias] %@", uid);
	[GeTuiSdk unbindAlias:uid andSequenceNum:@"1234567890"];
	
	return nil;
}

- (NSDictionary *)toClipboard:(NSDictionary * )args callId:(NSString *)callId{
    NSString * copyText = [args valueForKey:@"text"];
    NSLog(@"toClipboard:%@", copyText);
    
    UIPasteboard*pasteboard = [UIPasteboard generalPasteboard];
    pasteboard.string = copyText;
    
    return nil;
}

- (NSDictionary *)getDeviceUUID:(NSDictionary * )args callId:(NSString *)callId{
	return @{@"uuid": [Utils getUUIDWithMD5]};
}

- (NSDictionary *)navigateToUrl:(NSDictionary *)args callId:(NSString *)callId{
	NSString * url = [args valueForKey:@"url"];
	[[UIApplication sharedApplication] openURL:[NSURL URLWithString:url]];
	
	return nil;
}

- (NSDictionary*)uploadImageFromDevice:(NSDictionary *)args callId:(NSString *)callId{
	NSString * title = [args valueForKey:@"title"];
	NSString * uploadUrl = [args valueForKey:@"uploadUrl"];
	NSDictionary *options =@{
													 @"title": title,
													 @"cancelButtonTitle": @"取消",
													 @"takePhotoButtonTitle": @"拍照",
													 @"chooseFromLibraryButtonTitle": @"从相册中选取",
													 @"allowsEditing" : @YES,
													 @"maxWidth": @300,
													 @"maxHeight": @300
													 };
	[imagePicker show:options callback:^(NSMutableDictionary *response){
		if(response){
			NSURL *url = [NSURL URLWithString:uploadUrl];
			NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
			[request setHTTPMethod:@"POST"];
			
			NSData *data = response[@"data"];
			NSURLSessionUploadTask *uploadTask = [manager uploadTaskWithRequest:request fromData:data progress:nil completionHandler:^(NSURLResponse *response, id responseObject, NSError *error) {
    if (error) {
			//NSLog(@"Error: %@", error);
			[self returnCall:callId retArgs:@{@"result": @2}];
		} else {
			//NSLog(@"Success: %@ %@", response, responseObject);
			int code = [responseObject[@"code"] intValue];
			switch(code){
				case 0:
				{
					NSString *path = responseObject[@"data"][@"file_path"];
					[self returnCall:callId retArgs:@{@"result": @0, @"url": path}];
				}
					break;
				default:
					[self returnCall:callId retArgs:@{@"result": @3}];
					break;
			}
			
		}
			}];
			[uploadTask resume];
		}else{
			[self returnCall:callId retArgs:@{@"result": @1}];
		}
	}];
	
	return nil;
}

- (NSDictionary*)vibrate:(NSDictionary *)args callId:(NSString *)callId{
	AudioServicesPlaySystemSound(kSystemSoundID_Vibrate);
	
	return nil;
}

- (NSDictionary*)share:(NSDictionary *)args callId:(NSString *)callId{
    [[WXApiManager getInstance]doShare:args withCallback:^(int code) {
        [self returnCall:callId retArgs:@{@"code": @(code)}];
    }];
	
	return nil;
}

-(NSDictionary*)alert:(NSDictionary *)args callId:(NSString *)callId{
	NSString *title = args[@"title"];
	NSString *message = args[@"message"];
	NSArray *buttons = args[@"buttons"];
	
	UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
	
	for(NSString *button in buttons){
		UIAlertAction *action = [UIAlertAction actionWithTitle:button style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
			NSUInteger index = [buttons indexOfObject:[action title]];
			[self returnCall:callId retArgs:@{@"index": [NSNumber numberWithUnsignedInteger:index]}];
		}];
		[alertController addAction:action];
	}
	
	[viewController presentViewController:alertController animated:YES completion:nil];
	
	return nil;
}

- (NSDictionary*)addWebView:(NSDictionary *)args callId:(NSString *)callId{
	NSString * url = [args valueForKey:@"url"];
	
	[_webView addTo: viewController.view];
	[_webView loadByUrl:url];
	
	return nil;
}

- (NSDictionary*)removeWebView:(NSDictionary *)args callId:(NSString *)callId{
	[_webView remove];
	
	return nil;
}

- (NSDictionary*)authenticate:(NSDictionary *)args callId:(NSString *)callId{
    int type = [args[@"type"]intValue];
    if (type == 3) // 3 手机 1 通行证
    {
        [sdk showLoginByMobile];
    }
    if (type == 1) {
        [sdk showLoginByInput];
        
    }
    if (type == 2) {
        [sdk loginByGuest];
    }
	
    authCallId = callId;

	return nil;
}

- (NSDictionary*)authThirdPart:(NSDictionary *)args callId:(NSString *)callId{
	int type = [args[@"type"]intValue];
	switch(type){
		case 1: //weixin
           [[WXApiManager getInstance]doAuth:viewController withCallback:^(SendAuthResp *authResp) {
               [sdk onWXAuthCallback:authResp isDirect:YES];
           } withState:@"auth"];
			
			break;
	}
	
	authCallId = callId;
	
	return nil;
}

- (NSDictionary*)bindThirdPart:(NSDictionary *)args callId:(NSString *)callId{
	int type = [args[@"type"]intValue];
	switch(type){
		case 0: //mobile
			[sdk showBindByMobile];
			break;
		case 1: //weixin
                [[WXApiManager getInstance]doAuth:viewController withCallback:^(SendAuthResp *authResp) {
                    [sdk onWXBindCallback:authResp];
                } withState:@"auth"];
			break;
	}
	
	bindCallId = callId;
	
    
	return nil;
}
-(NSDictionary *)UMLoginStatistics:(NSDictionary *)args callId:(NSString *)callId{
    int type = [args[@"type"]intValue];
    NSString* uid = args[@"uid"];
    //NSString* struid = [uid stringvalue]
    if (type == 1) {
        [MobClick profileSignInWithPUID:uid provider:@"WX"];
    }
    else
    {
        [MobClick profileSignInWithPUID:uid];
    }
    
    return nil;
}
-(NSDictionary *)recharge:(NSDictionary *)args callId:(NSString *)callId{
	NSString *id = args[@"id"];
	int type = [args[@"type"]intValue];//0苹果 1 支付宝 2 微信
    // 默认type用1 此处直接在platform层修改，不影响原有native接口代码
    type = 2;
    int gold = [args[@"gold"]intValue];
    int money = [args[@"money"]intValue];
    int typeMob = 1;
    if (type == 1) {
        typeMob = 2;
    } else if (type == 2) {
        typeMob = 3;
    }
    [MobClickGameAnalytics pay:money source:typeMob coin:gold];
	//int type = 2;
	
	[appPay pay:id withType:type withCallback:^(int code) {
		[self returnCall:callId retArgs:@{@"code": @(code)}];
	}];
	
	return nil;
}

# pragma mark - AuthenticateDelegate

- (void)onAuthticateResult:(NSMutableDictionary *)result{
	[self returnCall:authCallId retArgs:result];
}

- (void)onAuthticateCancel{
	[self returnCall:authCallId retArgs:@{@"code": @(1)}];
}

# pragma mark - BindDelegate

- (void)onBindResult:(NSMutableDictionary *)result
				withUserInfo:(NSDictionary*)userInfo{
	if(userInfo){
		id data = result[@"data"];
		if([data isKindOfClass:[NSString class]]){
			data = [[NSMutableDictionary alloc]initWithCapacity:10];
		}
		for (NSString *key in userInfo) {
			[data setValue:userInfo[key] forKey:key];
		}
		[result setValue:data forKey:@"data"];
	}
	[self returnCall:bindCallId retArgs:result];
}

- (void)onBindCancel{
	[self returnCall:bindCallId retArgs:@{@"code": @(1)}];
}

- (void)submitLoginInfo{
    [[WebService getInstance] callApi:@"user" action:@"channel"
     params:@{@"channel_id": [Utils getInfo:@"InstallChannel"], @"uid": [GameInfo getInstance].uid, @"token": [GameInfo getInstance].token, @"os_type": @"ios"}
     callback:^(NSMutableDictionary *response) {
         if (response) {
             int code = [response[@"code"]intValue];
             if (code == 0) {
                 NSLog(@"submitLoginInfo success!");
             } else {
                 NSLog(@"submitLoginInfo failed! %@", response[@"code"]);
             }
         }
     }
     method: POST];
}


@end
